Feature/s65 113

LucUrlings opened the pull request.

Chazler committed:
CRUD added for adding a product. Only tests for Create done right now.

LucUrlings committed:
Added GetAll() method to get all products. Also updated a test to make it look better

Chazler committed:
Changed exceptions and added tests to the service

LucUrlings committed:
Merge branch 'feature/S65-113' of https://github.com/S65-2-project/MarketplaceService into feature/S65-113

LucUrlings committed:
Optimized imports and code cleanup over entire solution
Fixed the exception handling in the MarketplaceController.cs Update() method.
Added tests for the Controller and Repository. Optimized the Service tests.

LucUrlings committed:
Force rename to be able to rename it properly back

LucUrlings committed:
Fix the renaming problem

Chazler committed:
Fixed unittest

davydehaas98 committed:
Merge branch 'develop' into feature/S65-113

davydehaas98 commented on MarketplaceService/MarketplaceService.csproj:
```diff
@@ -4,6 +4,7 @@
         <TargetFramework>netcoreapp3.1</TargetFramework>
         <UserSecretsId>2145df81-9941-4fef-bc85-93ddde474a1f</UserSecretsId>
         <DockerDefaultTargetOS>Linux</DockerDefaultTargetOS>
+        <RootNamespace>marketplaceservice</RootNamespace>
```
Verander root namespace naar MarketplaceService

Chazler replied:
Dit geeft naming errorrs dus beter laten we het zo staan. Alternatief is om het hele project te renamen naar Microservice ipv Service

davydehaas98 replied:
Als ik het refactor krijg ik alleen een import issue als ik de service zelf aanroep. 
Dat kan je oplossen door dit bovenin de file te zetten zodat de juiste service wordt geimporteerd;
using MarketplaceService = MarketplaceService.Services.MarketplaceService;

davydehaas98 commented on MarketplaceService/Repositories/IMarketplaceRepository.cs:
```diff
@@ -0,0 +1,16 @@
+
+namespace marketplaceservice.Repositories
+{
+    public interface IMarketplaceRepository
```
XML doc op ALLE interface methodes

davydehaas98 commented on MarketplaceService/Services/IMarketplaceService.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace marketplaceservice.Services
+{
+    public interface IMarketplaceService
```
XML doc op ALLE interface methodes

davydehaas98 commented on MarketplaceService/Services/MarketplaceService.cs:
```diff
@@ -0,0 +1,57 @@
+{
+    public class MarketplaceService : IMarketplaceService
+    {
+        private readonly IMarketplaceRepository _repository;
```
_marketplaceRepository

davydehaas98 commented on MarketplaceService/Services/MarketplaceService.cs:
```diff
@@ -0,0 +1,57 @@
+        public async Task<Product> CreateProduct(CreateProductModel createProductModel)
+        {
+            if (string.IsNullOrEmpty(createProductModel.Title) || string.IsNullOrEmpty(createProductModel.Description))
+                throw new ArgumentException("All fields have to be filled out.");
```
Custom exceptions

davydehaas98 commented on MarketplaceService/Services/MarketplaceService.cs:
```diff
@@ -0,0 +1,57 @@
+        public async Task<Product> UpdateProduct(Guid id, UpdateProductModel updateProductModel)
+        {
+            if (string.IsNullOrEmpty(updateProductModel.Title) || string.IsNullOrEmpty(updateProductModel.Description))
+                throw new ArgumentException("All fields have to be filled out.");
```
Custom exceptions

davydehaas98 commented on MarketplaceService/Startup.cs:
```diff
@@ -25,27 +22,32 @@ public Startup(IConfiguration configuration)
         // This method gets called by the runtime. Use this method to add services to the container.
         public void ConfigureServices(IServiceCollection services)
```
Er moet nog authentication toegevoegd worden (JWT) hier en in de controllers.

Chazler replied:
Er is nog geen authenticatie methode voor derde services in de account service dus die staat nog on hold

davydehaas98 commented on MarketplaceService/appsettings.json:
```diff
@@ -0,0 +1,15 @@
+{
+  "MarketplaceDatabaseSettings": {
```
Deze fields moeten empty zijn.

davydehaas98 commented on MarketplaceServiceTests/Controller/MarketplaceControllerTests.cs:
```diff
@@ -0,0 +1,227 @@
+            };
+
+            _marketplaceService.Setup(x => x.CreateProduct(productModel))
+                .Throws<ArgumentException>();
```
Dit gaat falen zodra de argument exceptions weg zijn.

davydehaas98 commented on MarketplaceServiceTests/Controller/MarketplaceControllerTests.cs:
```diff
@@ -0,0 +1,227 @@
+            _marketplaceService.Setup(x => x.CreateProduct(createProductModel)).ReturnsAsync(product);
+
+            var result = await _marketplaceController.Create(createProductModel);
+            var data = result as ObjectResult;
```
Onnodige regel, zet as ObjectResult achter de await.

davydehaas98 commented on MarketplaceServiceTests/Controller/MarketplaceControllerTests.cs:
```diff
@@ -0,0 +1,227 @@
+
+            //Act
+            var result = await _marketplaceController.Get(product1.Id);
+            var data = result as ObjectResult;
```
Redundant

davydehaas98 commented on MarketplaceServiceTests/Controller/MarketplaceControllerTests.cs:
```diff
@@ -0,0 +1,227 @@
+
+            //Act
+            var result = await _marketplaceController.Update(guid, updateProductModel);
+            var data = result as ObjectResult;
```
Redundant. Zet `as ObjectResult` achter de result

davydehaas98 commented on MarketplaceServiceTests/MarketplaceServiceTests.csproj:
```diff
@@ -4,13 +4,21 @@
 
         <IsPackable>false</IsPackable>
+
+        <RootNamespace>marketplaceservicetests</RootNamespace>
```
Hier ook de namespace veranderen naar MarketplaceServiceTests

davydehaas98 commented on MarketplaceServiceTests/Repository/MarketplaceRepositoryTests.cs:
```diff
@@ -0,0 +1,178 @@
+            _mongoDbRunner.Dispose();
+        }
+
+        private readonly MongoDbRunner _mongoDbRunner;
```
Fields boven een constructor.

davydehaas98 commented on MarketplaceServiceTests/Repository/MarketplaceRepositoryTests.cs:
```diff
@@ -0,0 +1,178 @@
+
+        public void Dispose()
+        {
+            _mongoDbRunner.Dispose();
```
_mongoDbRunner?.Dispose();

davydehaas98 commented on MarketplaceServiceTests/Repository/MarketplaceRepositoryTests.cs:
```diff
@@ -0,0 +1,178 @@
+        private readonly IMarketplaceRepository _marketplaceRepository;
+
+        [Fact]
+        public async void CreateProduct()
```
Async kan niet een void returnen. Maak ipv van void hier overal Task van.

https://docs.microsoft.com/en-us/archive/msdn-magazine/2013/march/async-await-best-practices-in-asynchronous-programming

davydehaas98 commented on MarketplaceServiceTests/Repository/MarketplaceRepositoryTests.cs:
```diff
@@ -0,0 +1,178 @@
+
+            var result = await _marketplaceRepository.GetProduct(product1.Id);
+
+            // _testOutputHelper.WriteLine(result.Title);
```
Verwijder commented code.

davydehaas98 commented on MarketplaceServiceTests/Service/MarketplaceServiceTests.cs:
```diff
@@ -0,0 +1,151 @@
+            _marketplaceService = new MarketplaceService(_repository.Object);
+        }
+
+        private readonly IMarketplaceService _marketplaceService;
```
Fields boven constructor.

davydehaas98 commented on MarketplaceServiceTests/Repository/MarketplaceRepositoryTests.cs:
```diff
@@ -0,0 +1,178 @@
+
+namespace marketplaceservicetests.Repository
+{
+    // : IDisposable
```
Weghalen

Chazler replied:
Welke van de 4 lines? Als je een suggest gebruikt is het misschien iets duidelijker ;)

davydehaas98 commented on MarketplaceServiceTests/Service/MarketplaceServiceTests.cs:
```diff
@@ -0,0 +1,151 @@
+
+            _repository.Setup(x => x.CreateProduct(It.IsAny<Product>())).ReturnsAsync(product);
+
+            var result =
```
Nutteloze enter

davydehaas98 commented on MarketplaceServiceTests/Service/MarketplaceServiceTests.cs:
```diff
@@ -0,0 +1,151 @@
+            _repository.Setup(x => x.CreateProduct(It.IsAny<Product>())).ReturnsAsync(product);
+
+            var result =
+                await Assert.ThrowsAsync<ArgumentException>(() =>
```
Argument exception

davydehaas98 commented on MarketplaceServiceTests/Service/MarketplaceServiceTests.cs:
```diff
@@ -0,0 +1,151 @@
+            _repository.Setup(x =>
+                x.UpdateProduct(product.Id, product)).ReturnsAsync(updatedProduct);
+
+            var result = await Assert.ThrowsAsync<ArgumentException>(() =>
```
Argument Exception

davydehaas98 reviewed (changes requested):
⏰

Chazler committed:
Added review changes & custom exceptions

Chazler committed:
Added review changes & custom exceptions

Chazler committed:
Merge remote-tracking branch 'origin/feature/S65-113' into feature/S65-113

davydehaas98 commented on MarketplaceService/Exceptions/EmptyFieldException.cs:
```diff
@@ -0,0 +1,14 @@
+    public class EmptyFieldException : Exception
+    {
+        public EmptyFieldException()
+            : base("All fields have to be filled out.")
```
Filled out moet filled in zijn

Chazler replied:
https://www.linguee.nl/engels-nederlands/vertaling/needs+to+be+filled+out.html

davydehaas98 commented on MarketplaceService/Services/MarketplaceService.cs:
```diff
@@ -0,0 +1,58 @@
+    {
+        private readonly IMarketplaceRepository _marketplaceRepository;
+
+        public MarketplaceService(IMarketplaceRepository marketplaceMarketplaceRepository)
```
marketplaceMarketplaceRepository moet marketplaceRepository zijn

davydehaas98 commented on MarketplaceService/Repositories/IMarketplaceRepository.cs:
```diff
@@ -0,0 +1,40 @@
+        /// </summary>
+        /// <param name="productIn"></param>
+        /// <returns>Product</returns>
+        Task<Product> CreateProduct(Product productIn);
```
```suggestion
        Task<Product> CreateProduct(Product productIn);
        
```

davydehaas98 commented on MarketplaceService/Repositories/IMarketplaceRepository.cs:
```diff
@@ -0,0 +1,40 @@
+        /// Get all products
+        /// </summary>
+        /// <returns>Products</returns>
+        Task<List<Product>> GetAll();
```
```suggestion
        Task<List<Product>> GetAll();
        
```

davydehaas98 commented on MarketplaceService/Repositories/IMarketplaceRepository.cs:
```diff
@@ -0,0 +1,40 @@
+        /// </summary>
+        /// <param name="id"></param>
+        /// <returns>Product</returns>
+        Task<Product> GetProduct(Guid id);
```
```suggestion
        Task<Product> GetProduct(Guid id);
        
```

davydehaas98 commented on MarketplaceService/Repositories/IMarketplaceRepository.cs:
```diff
@@ -0,0 +1,40 @@
+        /// <param name="id"></param>
+        /// <param name="productIn"></param>
+        /// <returns>Product</returns>
+        Task<Product> UpdateProduct(Guid id, Product productIn);
```
```suggestion
        Task<Product> UpdateProduct(Guid id, Product productIn);
        
```

davydehaas98 commented on MarketplaceService/Services/MarketplaceService.cs:
```diff
@@ -0,0 +1,58 @@
+            if (string.IsNullOrEmpty(createProductModel.Title) || string.IsNullOrEmpty(createProductModel.Description))
+                throw new EmptyFieldException();
+
+            var productIn = new Product
```
Noem het product ipv productIn, het is namelijk geen parameter

davydehaas98 commented on MarketplaceService/Exceptions/EmptyFieldException.cs:
```diff
@@ -0,0 +1,14 @@
+        {
+        }
+    }
+
```
```suggestion

```

davydehaas98 commented on MarketplaceServiceTests/Controller/MarketplaceControllerTests.cs:
```diff
@@ -0,0 +1,224 @@
+{
+    public class MarketplaceControllerTests
+    {
+        public MarketplaceControllerTests()
```
Fields boven de constructor

davydehaas98 commented on MarketplaceServiceTests/Controller/MarketplaceControllerTests.cs:
```diff
@@ -0,0 +1,224 @@
+            //Arrange
+            var guid = Guid.NewGuid();
+
+            _marketplaceService.Setup(x => x.DeleteProduct(guid)).Throws<Exception>();
```
Hebben wij geen custom DeleteException die we throwen?

davydehaas98 commented on MarketplaceServiceTests/Controller/MarketplaceControllerTests.cs:
```diff
@@ -0,0 +1,224 @@
+                Description = descriptionText
+            };
+
+            _marketplaceService.Setup(x => x.GetProduct(product1.Id)).Throws<Exception>();
```
Hebben wij geen custom exception die we throwen als een get fout gaat?

davydehaas98 commented on MarketplaceServiceTests/Controller/MarketplaceControllerTests.cs:
```diff
@@ -0,0 +1,224 @@
+                Description = descriptionText
+            };
+
+            _marketplaceService.Setup(x => x.UpdateProduct(guid, updateProductModel)).Throws<Exception>();
```
Hebben wij geen custom exception die we throwen als een update fout gaat?

davydehaas98 commented on MarketplaceServiceTests/Repository/MarketplaceRepositoryTests.cs:
```diff
@@ -0,0 +1,178 @@
+
+namespace marketplaceservicetests.Repository
+{
+
```
```suggestion
```

davydehaas98 commented on MarketplaceServiceTests/Repository/MarketplaceRepositoryTests.cs:
```diff
@@ -0,0 +1,178 @@
+
+    public class MarketplaceRepositoryTests : IDisposable
+    {
+        
```
```suggestion
```

davydehaas98 commented on MarketplaceServiceTests/Service/MarketplaceServiceTests.cs:
```diff
@@ -0,0 +1,151 @@
+    public class MarketplaceServiceTests
+    {
+        private readonly IMarketplaceService _marketplaceService;
+        private readonly Mock<IMarketplaceRepository> _repository;
```
Rename _repository naar _marketplaceRepository voor consistentie

davydehaas98 reviewed (changes requested):
Namespace is nog lowercase

Chazler committed:
S65-133: Update MarketplaceService/Repositories/IMarketplaceRepository.cs

Co-authored-by: Davy de Haas <davy.dehaas98@gmail.com>

Chazler committed:
S65-133: Update MarketplaceService/Repositories/IMarketplaceRepository.cs

Co-authored-by: Davy de Haas <davy.dehaas98@gmail.com>

Chazler committed:
S65-133: Update MarketplaceService/Repositories/IMarketplaceRepository.cs

Co-authored-by: Davy de Haas <davy.dehaas98@gmail.com>

Chazler committed:
S65-133: Update MarketplaceServiceTests/Repository/MarketplaceRepositoryTests.cs

Co-authored-by: Davy de Haas <davy.dehaas98@gmail.com>

Chazler committed:
S65-133: Update MarketplaceServiceTests/Repository/MarketplaceRepositoryTests.cs

Co-authored-by: Davy de Haas <davy.dehaas98@gmail.com>

Chazler committed:
S65-133: Update MarketplaceService/Repositories/IMarketplaceRepository.cs

Co-authored-by: Davy de Haas <davy.dehaas98@gmail.com>

Chazler committed:
S65-113: Update MarketplaceService/Exceptions/EmptyFieldException.cs

Co-authored-by: Davy de Haas <davy.dehaas98@gmail.com>

Chazler committed:
S65-113: Added requested changes. Created extra custom exceptions and changed namespace.

Chazler committed:
Merge remote-tracking branch 'origin/feature/S65-113' into feature/S65-113

davydehaas98 commented on MarketplaceService/Exceptions/ProductNotFoundException.cs:
```diff
@@ -0,0 +1,13 @@
+    public class ProductNotFoundException: Exception
+    {
+        public ProductNotFoundException()
+            : base("All fields have to be filled out.")
```
Ik neem aan dat dit vergeten was met de copy paste om aan te passen naar product was not found of iets dergelijks :)

Chazler replied:
Oh derp, ja dat klopt.

davydehaas98 reviewed (changes requested):
Een exception message veranderen en dan is het prima.

Chazler committed:
S65-113: Changed productnotfound exception message

davydehaas98 reviewed (approved):
🥇
